app/src/main/java/com/github/se/gatherspot/ProfileFirebaseConnection.kt

```diff
@@ -3,77 +3,154 @@ package com.github.se.gatherspot
     val data =
         hashMapOf(
-            "userName" to profile.userName,
-            "bio" to profile.bio,
-            "image" to profile.image,
-            "interests" to Interests.toCompressedString(profile.interests))
-    db.collection(tag)
-        .document(profile.id)
+            "userName" to element.userName,
+            "bio" to element.bio,
+            "image" to element.image,
+            "interests" to Interests.toCompressedString(element.interests))
+    Firebase.firestore
+        .collection(COLLECTION)
+        .document(element.id)
         .set(data)
-        .addOnSuccessListener { Log.d(tag, "DocumentSnapshot successfully written!") }
-        .addOnFailureListener { e -> Log.w(tag, "Error writing document", e) }
+        .addOnSuccessListener { Log.d(TAG, "DocumentSnapshot successfully written!") }
+        .addOnFailureListener { e -> Log.w(TAG, "Error writing document", e) }
   }
 
-  /**
-   * Deletes the profile from the database
-   *
-   * @param id the id of the profile to delete
-   */
-  fun deleteFromFirebase(id: String) {
-    // TODO : WE WILL NEED TO REMOVE EVERYTHING RELATED TO THE USER, LIKE REGISTERED_EVENTS, etc...
-    // probably want to do that at a later date in the project, when the picture is complete
-    db.collection(tag)
+  override fun update(id: String, field: String, value: Any) {
+    when (field) {
+      "interests" -> {
+        when (value) {
+          is Set<*> -> {
+            updateInterests(id, value as Set<Interests>)
+            return
+          }
+          is List<*> -> { // This is already a misuse case please don't land here
+            Log.d(TAG, "Please use a Set instead of a List when updating interests of a profile")
+            updateInterests(id, value.toSet() as Set<Interests>)
+            return
+          }
+          is String -> {
+            super.update(id, field, value)
+            return
+          }
+        }
+      }
+      "userName" -> {
+        ifUsernameExists(value as String) { exists ->
+          if (exists) {
+            Log.d(TAG, "Username already exists")
+            return@ifUsernameExists
+          }
+        }
+      }
+    }
+
+    super.update(id, field, value)
+  }
+
+  fun update(profile: Profile) {
+    this.add(profile)
+  }
+
+  fun updateInterests(id: String, interests: Set<Interests>) {
+    Firebase.firestore
+        .collection(TAG)
         .document(id)
-        .delete()
-        .addOnSuccessListener { Log.d(tag, "DocumentSnapshot successfully deleted!") }
-        .addOnFailureListener { e -> Log.w(tag, "Error deleting document", e) }
+        .update("interests", Interests.toCompressedString(interests))
+        .addOnSuccessListener { Log.d(TAG, "DocumentSnapshot successfully updated!") }
+        .addOnFailureListener { e -> Log.w(TAG, "Error updating document", e) }
+  }
+
+  override fun delete(id: String) {
+    // delete associated data from other collection TODO
```

Comment:
No not yet it will be the subject of another task. We need to do a "deepsearch" to find other documents related to the profile in other collections and delete or update them.